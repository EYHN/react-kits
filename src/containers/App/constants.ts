import { appThemes } from '../../withStyles';
import { translationMessages } from '../../i18n';

export const DEFAULT_LOCALE: keyof typeof translationMessages = 'en';

export const DEFAULT_THEMENAME: keyof typeof appThemes = 'light';
