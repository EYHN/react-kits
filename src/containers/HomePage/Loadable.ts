import LoadingIndicator from '../../components/LoadingIndicator';
const Loadable = require('react-loadable');

export default Loadable({
  loader: () => import('./index'),
  loading: LoadingIndicator
});
