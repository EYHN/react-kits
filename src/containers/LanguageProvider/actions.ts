import {
  CHANGE_LOCALE,
} from './constants';

export interface IAction {
  type: typeof CHANGE_LOCALE;
  locale: string;
}

export function changeLocale(languageLocale: string) {
  return { type: CHANGE_LOCALE as typeof CHANGE_LOCALE, locale: languageLocale };
}
