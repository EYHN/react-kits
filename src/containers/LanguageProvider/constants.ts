export const CHANGE_LOCALE = 'app/LanguageToggle/CHANGE_LOCALE';
export const LOAD_LOCALE_SUCCESS = 'app/LanguageToggle/LOAD_LOCALE_SUCCESS';
export const LOAD_LOCALE_ERROR = 'app/LanguageToggle/LOAD_LOCALE_ERROR';
