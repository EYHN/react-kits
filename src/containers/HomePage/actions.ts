import { LOAD_HITOKOTO, LOAD_HITOKOTO_SUCCESS, LOAD_HITOKOTO_ERROR } from 'containers/HomePage/constants';

export interface ILoadHitokotoAction {
  type: typeof LOAD_HITOKOTO;
}

export interface IHitokotoLoaded {
  type: typeof LOAD_HITOKOTO_SUCCESS;
  hitokoto: string;
}

export interface IHitokotoLoadingError {
  type: typeof LOAD_HITOKOTO_ERROR;
  error: any;
}

export function loadHitokoto(): ILoadHitokotoAction {
  return {
    type: LOAD_HITOKOTO,
  };
}

export function hitokotoLoaded(hitokoto: string): IHitokotoLoaded {
  return {
    type: LOAD_HITOKOTO_SUCCESS,
    hitokoto
  };
}

export function hitokotoLoadingError(error: any): IHitokotoLoadingError {
  return {
    type: LOAD_HITOKOTO_ERROR,
    error
  };
}
