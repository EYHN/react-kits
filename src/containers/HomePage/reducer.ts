import { fromJS } from 'immutable';

import {
  LOAD_HITOKOTO,
  LOAD_HITOKOTO_SUCCESS,
  LOAD_HITOKOTO_ERROR
} from './constants';
import { ILoadHitokotoAction, IHitokotoLoaded, IHitokotoLoadingError } from 'containers/HomePage/actions';

const initialState = fromJS({
  loading: false,
  error: false,
  hitokoto: undefined
});

type IAction = ILoadHitokotoAction | IHitokotoLoaded | IHitokotoLoadingError;

export default function homeReducer(state = initialState, action: IAction) {
  switch (action.type) {
    case LOAD_HITOKOTO:
      return state
        .set('loading', true)
        .set('error', false)
        .set('hitokoto', undefined);
    case LOAD_HITOKOTO_SUCCESS:
      return state
        .set('loading', false)
        .set('hitokoto', action.hitokoto);
    case LOAD_HITOKOTO_ERROR:
      return state
        .set('error', action.error)
        .set('loading', false);
    default:
      return state;
  }
}
