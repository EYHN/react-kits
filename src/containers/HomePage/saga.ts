import { hitokotoLoaded, hitokotoLoadingError, loadHitokoto } from 'containers/HomePage/actions';
import { put, takeLatest, call } from 'redux-saga/effects';
import { LOAD_HITOKOTO } from 'containers/HomePage/constants';

export function* getHitokoto() {
  try {
    const res = yield call(fetch, 'https://sslapi.hitokoto.cn/');
    const data = yield call(res.json);
    const hitokoto = yield data.hitokoto;
    yield put(hitokotoLoaded(hitokoto));
  } catch (err) {
    yield put(hitokotoLoadingError(err));
  }
}

export default function* hitokotoData() {
  yield takeLatest(LOAD_HITOKOTO, getHitokoto);
  yield put(loadHitokoto());
}
