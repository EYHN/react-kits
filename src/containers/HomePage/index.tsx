import { returntypeof } from 'react-redux-typescript';
import { createSelector } from 'reselect';
import React from 'react';
import { Dispatch, compose } from 'redux';
import { connect } from 'react-redux';
import Helloworld from 'components/helloworld/index';
import { changeLocale } from 'containers/LanguageProvider/actions';
import { makeSelectLocale } from 'containers/LanguageProvider/selectors';
import { appLocales } from 'i18n';
import { changeTheme } from 'containers/ThemeProvider/actions';
import { makeSelectThemeName } from 'containers/ThemeProvider/selectors';
import { appThemes } from 'withStyles';
import { loadHitokoto } from 'containers/HomePage/actions';
import reducer from './reducer';
import saga from './saga';

import injectReducer from 'utils/injectReducer';
import injectSaga from 'utils/injectSaga';
import { makeSelectHitokoto } from 'containers/HomePage/selectors';

interface IHomePageProps {
}

const mapStateToProps = createSelector(
  makeSelectLocale(),
  makeSelectThemeName(),
  makeSelectHitokoto(),
  (locale, theme, hitokoto) => ({ locale, theme, hitokoto })
);

export const mapDispatchToProps = (dispatch: Dispatch<{}>) => ({
  changeLocale: (locale: string) => { dispatch(changeLocale(locale)); },
  changeTheme: (theme: string) => {dispatch(changeTheme(theme)); },
  onGetHitokoto: () => (dispatch(loadHitokoto()))
});

const stateProps = returntypeof(mapStateToProps);
const dispatchProps = returntypeof(mapDispatchToProps);

type Props = typeof stateProps & IHomePageProps & typeof dispatchProps;

export class HomePage extends React.PureComponent<Props, undefined> {

  handleLocaleSelectChange: React.ReactEventHandler<HTMLSelectElement> = (e) => {
    this.props.changeLocale(e.currentTarget.value);
  }

  handleThemeSelectChange: React.ReactEventHandler<HTMLSelectElement> = (e) => {
    this.props.changeTheme(e.currentTarget.value);
  }

  public render() {
    const appLocalesItems = appLocales.map((locale) => {
      return <option key={locale} value={locale}>{locale}</option>;
    });
    const appThemesItems = Object.keys(appThemes).map((theme) => {
      return <option key={theme} value={theme}>{theme}</option>;
    });
    return (
      <div>
        <Helloworld />
        <select
          name='Locale'
          value={this.props.locale}
          onChange={this.handleLocaleSelectChange}
        >
          {appLocalesItems}
        </select>
        <select
          name='Theme'
          value={this.props.theme}
          onChange={this.handleThemeSelectChange}
        >
          {appThemesItems}
        </select>
        <p>{this.props.hitokoto}</p>
      </div>
    );
  }
}

// tslint:disable-next-line:max-line-length
const withConnect = connect<typeof stateProps, typeof dispatchProps, IHomePageProps>(mapStateToProps, mapDispatchToProps);

const withReducer = injectReducer({ key: 'home', reducer });
const withSaga = injectSaga({ key: 'home', saga });

export default compose(
  withReducer,
  withSaga,
  withConnect
)(HomePage);
