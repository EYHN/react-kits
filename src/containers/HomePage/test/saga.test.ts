import hitokotoData, {getHitokoto} from 'containers/HomePage/saga';
import {put, takeLatest} from 'redux-saga/effects';
import {hitokotoLoaded, hitokotoLoadingError, loadHitokoto} from 'containers/HomePage/actions';
import {LOAD_HITOKOTO} from 'containers/HomePage/constants';

describe('getHitokoto Saga', () => {
  let getHitokotoGenerator: IterableIterator<any>;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getHitokotoGenerator = getHitokoto();

    const callfetchDescriptor = getHitokotoGenerator.next().value;
    expect(callfetchDescriptor).toMatchSnapshot();

    const calljsonDescriptor = getHitokotoGenerator.next({json: () => ({})});
    expect(calljsonDescriptor).toMatchSnapshot();

    const getHitokotoDescriptor = getHitokotoGenerator.next({hitokoto: '123'});
    expect(getHitokotoDescriptor).toMatchSnapshot();
  });

  it('should dispatch the hitokotoLoaded action if it requests the data successfully', () => {
    const hitokoto = '123';
    const putDescriptor = getHitokotoGenerator.next(hitokoto).value;
    expect(putDescriptor).toEqual(put(hitokotoLoaded(hitokoto)));
  });

  it('should call the hitokotoLoadingError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getHitokotoGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(hitokotoLoadingError(response)));
  });
});

describe('hitokotoData Saga', () => {
  let hitokotoDataSaga: IterableIterator<any>;

  beforeEach(() => {
    hitokotoDataSaga = hitokotoData();
  });

  it('should start task to watch for LOAD_HITOKOTO action', () => {
    const takeLatestDescriptor = hitokotoDataSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(LOAD_HITOKOTO, getHitokoto));
  });

  it('should put LOAD_HITOKOTO action', () => {
    hitokotoDataSaga.next();
    const putDescriptor = hitokotoDataSaga.next().value;
    expect(putDescriptor).toEqual(put(loadHitokoto()));
  });
});
