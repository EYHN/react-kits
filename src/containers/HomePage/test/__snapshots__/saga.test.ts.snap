// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getHitokoto Saga should call the hitokotoLoadingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://sslapi.hitokoto.cn/",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`getHitokoto Saga should call the hitokotoLoadingError action if the response errors 2`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "CALL": Object {
      "args": Array [],
      "context": null,
      "fn": [Function],
    },
  },
}
`;

exports[`getHitokoto Saga should call the hitokotoLoadingError action if the response errors 3`] = `
Object {
  "done": false,
  "value": "123",
}
`;

exports[`getHitokoto Saga should dispatch the hitokotoLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://sslapi.hitokoto.cn/",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`getHitokoto Saga should dispatch the hitokotoLoaded action if it requests the data successfully 2`] = `
Object {
  "done": false,
  "value": Object {
    "@@redux-saga/IO": true,
    "CALL": Object {
      "args": Array [],
      "context": null,
      "fn": [Function],
    },
  },
}
`;

exports[`getHitokoto Saga should dispatch the hitokotoLoaded action if it requests the data successfully 3`] = `
Object {
  "done": false,
  "value": "123",
}
`;
